### PR TITLE
num: add Abs() tests

### PR DESCRIPTION
**Pull request by nisfornature**
This change adds supporting test cases for absolute value functions, Abs(), in num/dual and num/dualcmplx.

### CODE REVIEW

File: num/dual/dual_test.go [hunk 1]
```diff
@@ -387,6 +387,74 @@ func TestPowReal(t *testing.T) {
...
+	{d: Number{Real: 1, Emag: 0}, want: Number{Real: 1, Emag: 0}},
+	{d: Number{Real: 1, Emag: 1}, want: Number{Real: 1, Emag: 1}},
+	{d: Number{Real: 1, Emag: 2}, want: Number{Real: 1, Emag: 2}},
+	{d: Number{Real: 1, Emag: 3}, want: Number{Real: 1, Emag: 3}},
+	{d: Number{Real: 1, Emag: math.NaN()}, want: Number{Real: 1, Emag: math.NaN()}},
+	{d: Number{Real: 1, Emag: math.Inf(-1)}, want: Number{Real: 1, Emag: math.Inf(-1)}},
+	{d: Number{Real: 1, Emag: math.Inf(1)}, want: Number{Real: 1, Emag: math.Inf(1)}},
+
+	// AbsReal(Infinity+xϵ) = 0+NaNϵ for |x| < 1
+	{d: Number{Real: math.Inf(1), Emag: 0}, want: Number{Real: math.Inf(1), Emag: 0}},
```
Comment (kortschak) [inline num/dual/dual_test.go]: These don't match the comment.

File: num/dual/dual_test.go [hunk 2]
```diff
@@ -387,6 +387,74 @@ func TestPowReal(t *testing.T) {
...
+	{d: Number{Real: math.Inf(1), Emag: 0.5}, want: Number{Real: math.Inf(1), Emag: 0.5}},
+
+	// AbsReal(±x+yϵ) = |x|+NaNϵ for 0 < |x| < 1
+	{d: Number{Real: 1.175495e-38, Emag: math.NaN()}, want: Number{Real:1.175495e-38, Emag:math.NaN()}},
+	{d: Number{Real: -1.175495e-38, Emag: math.NaN()}, want: Number{Real:1.175495e-38, Emag:math.NaN()}},
+	{d: Number{Real: -0.2, Emag: math.NaN()}, want: Number{Real: 0.2, Emag: math.NaN()}},
+	{d: Number{Real: -0.1, Emag: math.NaN()}, want: Number{Real: 0.1, Emag: math.NaN()}},
+	{d: Number{Real: 0.1, Emag: math.NaN()}, want: Number{Real: 0.1, Emag: math.NaN()}},
+	{d: Number{Real: 0.2, Emag: math.NaN()}, want: Number{Real: 0.2, Emag: math.NaN()}},
+
```
Comment (kortschak) [inline num/dual/dual_test.go]: Please delete this blank line.

File: num/dualcmplx/dual_test.go [hunk 3]
```diff
@@ -441,6 +441,65 @@ func TestPowReal(t *testing.T) {
...
+	d    Number
+	want float64
+}{
+	// Abs(NaN+NaNϵ) = NaN
+	{d: Number{Real: cmplx.NaN(), Dual: cmplx.NaN()}, want: math.NaN()},
+	{d: Number{Real: complex(math.NaN(), 10), Dual: 0+3i}, want: math.NaN()},
+	{d: Number{Real: complex(0, math.NaN()), Dual: 0+1i}, want: math.NaN()},
+
+	// Abs(±Infinity+NaNϵ) = +Infinity
+	{d: Number{Real: cmplx.Inf(), Dual: 0}, want: math.Inf(1)},
```
Comment (kortschak) [inline num/dualcmplx/dual_test.go]: This is not a test for the comment.

File: num/dualcmplx/dual_test.go [hunk 4]
```diff
@@ -441,6 +441,65 @@ func TestPowReal(t *testing.T) {
...
+	{d: Number{Real: -1+-1i, Dual: cmplx.NaN()}, want: math.Sqrt(2)},
+	{d: Number{Real: -1+1i, Dual: cmplx.NaN()}, want: math.Sqrt(2)},
+	{d: Number{Real: 1+-1i, Dual: cmplx.NaN()}, want: math.Sqrt(2)},
+	{d: Number{Real: -1+0i, Dual: cmplx.NaN()}, want: 1},
+	{d: Number{Real: 0+-1i, Dual: cmplx.NaN()}, want: 1},
+	{d: Number{Real: 1i, Dual: -3i}, want: 1},
+	{d: Number{Real: -1i, Dual: 2i}, want: 1},
+	{d: Number{Real: 1, Dual: 3i}, want: 1},
+	{d: Number{Real: -1, Dual: cmplx.NaN()}, want: 1},
+
```
Comment (kortschak) [inline num/dualcmplx/dual_test.go]: Please delete this blank line.

File: num/dualcmplx/dual_test.go [hunk 5]
```diff
@@ -441,6 +441,65 @@ func TestPowReal(t *testing.T) {
...
+	{d: Number{Real: complex(-1.175495e-38,0), Dual: cmplx.NaN()}, want: 1.175495e-38},
+	{d: Number{Real: -0.2i, Dual: cmplx.NaN()}, want: 0.2},
+	{d: Number{Real: -0.1i, Dual: cmplx.NaN()}, want: .1},
+}
+
+func TestAbsCmplx(t *testing.T) {
+	const tol = 1e-15
+	for _, test := range absCmplxTests {
+		result := Abs(test.d)
+		isExpectedAbsoluteValue := (math.IsNaN(result) && math.IsNaN(test.want)) || floats.EqualWithinAbsOrRel(result, test.want, tol, tol)
```
Comment (kortschak) [inline num/dualcmplx/dual_test.go]: Please write a helper for this, rather than using an intermediate variable.

File: num/dualcmplx/dual_test.go [hunk 6]
```diff
@@ -441,6 +441,73 @@ func TestPowReal(t *testing.T) {
...
+	{d: Number{Real: -0.2i, Dual: cmplx.NaN()}, want: 0.2},
+	{d: Number{Real: -0.1i, Dual: cmplx.NaN()}, want: .1},
+
+	// Abs(x) =  Sqrt( Sqrt(real(x)) + Sqrt(imaginary(x)) for any complex number x
+	{d: Number{Real: 1i, Dual: -3i}, want: 1},
+	{d: Number{Real: -1i, Dual: 2i}, want: 1},
+	{d: Number{Real: 1, Dual: 3i}, want: 1},
+}
+
+func Complex128ToNumber(val complex128) Number {
```
Comment (golangcibot) [inline num/dualcmplx/dual_test.go]: U1000: func `Complex128ToNumber` is unused (from `unused`)

File: optimize/gradientdescent.go [hunk 7]
```diff
@@ -48,6 +48,7 @@ func (g *GradientDescent) Init(dim, tasks int) int {
 func (g *GradientDescent) Run(operation chan<- Task, result <-chan Task, tasks []Task) {
 	g.status, g.err = localOptimizer{}.run(g, g.GradStopThreshold, operation, result, tasks)
 	close(operation)
+	return
```
Comment (golangcibot) [inline optimize/gradientdescent.go]: S1023: redundant `return` statement (from `gosimple`)

File: num/dual/dual_test.go (hunk 2 above)
Comment (nisfornature) [reply num/dual/dual_test.go]: > Please squash out the unnecessary commits here and remove the vim swap files.
> 
> It would be helpful if you provided the sources of the special case claims you are making in the tests.

Sorry about that. Thanks for your time. My intention is to add test coverage for functions discussed in #1093.

I just wanted to include edge and corner cases.

File: num/dualcmplx/dual_test.go (hunk 3 above)
Comment (nisfornature) [reply num/dualcmplx/dual_test.go]: Thanks for your keen eye!

File: num/dual/dual_test.go (hunk 2 above)
Comment (kortschak) [reply num/dual/dual_test.go]: We still need provenance. The current code matches exactly the reference implementation for dual and is trivially correct for the dual complex according to the reference paper.
